src/main/flight/gps_rescue.c

```diff
@@ -0,0 +1,373 @@
+    }
+}
+
+void rescueStart()
+{
+    rescueState.phase = RESCUE_INITIALIZE;
+}
+
+void rescueStop()
+{
+    rescueState.phase = RESCUE_IDLE;
+}
+
+// Things that need to run regardless of GPS rescue mode being enabled or not
+void idleTasks()
+{
+    gpsRescueAngle[AI_PITCH] = 0;
+    gpsRescueAngle[AI_ROLL] = 0;
+
+    // Store the max altitude we see not during RTH so we know our fly-back minimum alt
+    rescueState.sensor.maxAltitude = MAX(rescueState.sensor.currentAltitude, rescueState.sensor.maxAltitude);
+    // Store the max distance to home during normal flight so we know if a flyaway is happening
+    rescueState.sensor.maxDistanceToHome = MAX(rescueState.sensor.distanceToHome, rescueState.sensor.maxDistanceToHome);
+
+    rescueThrottle = rcCommand[THROTTLE];
+
+    //to do: have a default value for hoverThrottle
+    float ct = getCosTiltAngle();
+    if (ct > 0.5 && ct < 0.96 && throttleSamples < 1E6 && rescueThrottle > 1070) { //5 to 45 degrees tilt
+        //TO DO: only sample when acceleration is low
+        uint16_t adjustedThrottle = 1000 + (rescueThrottle - 1000) * ct;
+        if (throttleSamples == 0) {
+            averageThrottle = adjustedThrottle;
+        } else {
+            averageThrottle += (adjustedThrottle - averageThrottle) / (throttleSamples + 1);
+        }
+        hoverThrottle = lrintf(averageThrottle);
+        throttleSamples++;
+    }
+
+}
+
+void rescueAttainPosition()
+{
+    // Point to home if that is in our intent
+    if (rescueState.intent.crosstrack) {
+        setBearing(rescueState.sensor.directionToHome);
+    }
+
+    if (!newGPSData) {
+        return;
+    }
+
+    /**
+        Speed controller
+    */
+    static float previousSpeedError = 0;
+    static int16_t speedIntegral = 0;
+
+    const int16_t speedError = (rescueState.intent.targetGroundspeed - rescueState.sensor.groundSpeed) / 100;
+    const int16_t speedDerivative = speedError - previousSpeedError;
+
+    speedIntegral = constrain(speedIntegral + speedError, -100, 100);
+
+    previousSpeedError = speedError;
+
+    int16_t angleAdjustment =  gpsRescue()->vP * speedError + (gpsRescue()->vI * speedIntegral) / 100 +  gpsRescue()->vD * speedDerivative;
+
+    gpsRescueAngle[AI_PITCH] = constrain(gpsRescueAngle[AI_PITCH] + MIN(angleAdjustment, 80), rescueState.intent.minAngle * 100, rescueState.intent.maxAngle * 100);
+
+    float ct = cos(DECIDEGREES_TO_RADIANS(gpsRescueAngle[AI_PITCH] / 10));
+
+    /**
+        Altitude controller
+    */
+    static float previousAltitudeError = 0;
+    static int16_t altitudeIntegral = 0;
+
+    const int16_t altitudeError = (rescueState.intent.targetAltitude - rescueState.sensor.currentAltitude) / 100; // Error in meters
+    const int16_t altitudeDerivative = altitudeError - previousAltitudeError;
+
+    // Only allow integral windup within +-15m absolute altitude error
+    if (ABS(altitudeError) < 25) {
+        altitudeIntegral = constrain(altitudeIntegral + altitudeError, -250, 250);
```

Comment:
What if [-250;250] is not enough to sustain altitude?

Reply:
I suppose it is possible that it could result in an inability to gain altitude, but it would have to be misconfigured pretty badly.  We are going to implement your idea of square root extraction on it so we can allow higher windup, but we did not want to have to re-tune the default gains prior to merging this with master, since finding a good one size fits all tune took a bit of work on @dbasch 's part.